src/System.Resources.Reader/src/System/Resources/ResourceReader.cs

```diff
@@ -50,31 +45,31 @@ public void Dispose()
         [System.Security.SecuritySafeCritical]  // auto-generated
         private unsafe void Dispose(bool disposing)
         {
-            if (_store != null)
+            if (_stream != null)
             {
 
                 if (disposing)
                 {
                     // Close the stream in a thread-safe way.  This fix means 
                     // that we may call Close n times, but that's safe.
-                    BinaryReader copyOfStore = _store;
-                    _store = null;
-                    if (copyOfStore != null)
-                        copyOfStore.Dispose();
+                    Stream stream = _stream;
+                    _stream = null;
+                    if (stream != null)
+                        stream.Dispose();
                 }
-                _store = null;
+                _stream = null;
                 _namePositions = null;
 
             }
         }
         private void SkipString()
         {
-            int stringLength = Read7BitEncodedInt();
-            if (stringLength < 0)
+            int stringLength;
+            if (!_stream.TryReadInt327BitEncoded(out stringLength) || stringLength < 0)
```

Comment:
It looks like these changes (here and other places reading is done) could be causing different exception types to be thrown.  For example, when BinaryReader can't fill its buffer with the requested number of bytes, it throws an EndOfStreamException.